File: android/app/src/main/java/net/wellvo/android/data/models/Family.kt
package net.wellvo.android.data.models

import androidx.compose.runtime.Immutable
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Immutable
@Serializable
data class Family(
    val id: String,
    val name: String,
    @SerialName("owner_id")
    val ownerId: String,
    @SerialName("subscription_tier")
    val subscriptionTier: SubscriptionTier,
    @SerialName("subscription_status")
    val subscriptionStatus: SubscriptionStatus,
    @SerialName("subscription_expires_at")
    val subscriptionExpiresAt: String? = null,
    // Grandfathering deadline for legacy Free-tier families. When set and in
    // the past, clients should gate paid features and prompt the Owner to
    // upgrade to Caregiver. NULL for families created after the Caregiver
    // tier migration.
    @SerialName("free_tier_expires_at")
    val freeTierExpiresAt: String? = null,
    @SerialName("max_receivers")
    val maxReceivers: Int,
    @SerialName("max_viewers")
    val maxViewers: Int,
    @SerialName("created_at")
    val createdAt: String
)

@Serializable
enum class SubscriptionTier {
    // Legacy tier, kept for grandfathered families created before the
    // Caregiver tier launched. New signups never land here.
    @SerialName("free") Free,
    // Lowest paid tier, sized for 1 Receiver + 3 Viewers (the dementia-
    // caregiver persona). $3.99/mo or $29.99/yr.
    @SerialName("caregiver") Caregiver,
    @SerialName("family") Family,
    @SerialName("family_plus") FamilyPlus
}

@Serializable
enum class SubscriptionStatus {
    @SerialName("active") Active,
    @SerialName("expired") Expired,
    @SerialName("grace_period") GracePeriod,
    @SerialName("cancelled") Cancelled
}

@Immutable
@Serializable
data class FamilyMember(
    val id: String,
    @SerialName("family_id")
    val familyId: String,
    @SerialName("user_id")
    val userId: String,
    val role: UserRole,
    val status: MemberStatus,
    @SerialName("invited_at")
    val invitedAt: String? = null,
    @SerialName("joined_at")
    val joinedAt: String? = null,
    @SerialName("users")
    val user: AppUser? = null
)

File: ios/DailyOK/Models/Family.swift
import Foundation

enum SubscriptionTier: String, Codable {
    /// Legacy tier, kept for grandfathered families created before the
    /// Caregiver tier launched. New signups never land here.
    case free
    /// Lowest paid tier, sized for 1 Receiver + 3 Viewers (the dementia-
    /// caregiver persona). $3.99/mo or $29.99/yr.
    case caregiver
    case family
    case familyPlus = "family_plus"

    // Forgiving decode (backward-compat rule): a tier this build doesn't know —
    // a future paid tier returned to an older app — must NOT throw the whole
    // `Family` decode. That previously took down the owner dashboard and the
    // receiver's status/check-in for every installed older build. Unknown →
    // `.free`: real feature gating uses the StoreKit-derived
    // `SubscriptionService.currentTier`, and a future-tier family carries no
    // free-tier grandfather deadline, so nothing is wrongly gated.
    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = SubscriptionTier(rawValue: raw) ?? .free
    }
}

enum SubscriptionStatus: String, Codable {
    case active
    case expired
    case gracePeriod = "grace_period"
    case cancelled

    // Forgiving decode: an unknown status (e.g. a future "paused"/"trial") must
    // not throw the `Family` decode. Unknown → `.active` (nothing gates on this
    // field today; the neutral choice avoids a false expiry banner).
    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = SubscriptionStatus(rawValue: raw) ?? .active
    }
}

struct Family: Codable, Identifiable {
    let id: UUID
    var name: String
    let ownerId: UUID
    var subscriptionTier: SubscriptionTier
    var subscriptionStatus: SubscriptionStatus
    var subscriptionExpiresAt: Date?
    /// Grandfathering deadline for legacy Free-tier families. When set and in
    /// the past, clients should gate paid features and prompt the Owner to
    /// upgrade to Caregiver. NULL for families created after the Caregiver
    /// tier migration.
    var freeTierExpiresAt: Date?
    var maxReceivers: Int
    var maxViewers: Int
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, name
        case ownerId = "owner_id"
        case subscriptionTier = "subscription_tier"
        case subscriptionStatus = "subscription_status"
        case subscriptionExpiresAt = "subscription_expires_at"
        case freeTierExpiresAt = "free_tier_expires_at"
        case maxReceivers = "max_receivers"
        case maxViewers = "max_viewers"
        case createdAt = "created_at"
    }

    /// True once a grandfathered Free-tier family's free window has lapsed
    /// (migration 00019 sets a 90-day deadline). Paid features should be gated
    /// and an upgrade prompt shown past this point (US-IOS097). A nil deadline
    /// means no grandfather window applies (so never "expired").
    var isFreeTierExpired: Bool {
        guard subscriptionTier == .free, let deadline = freeTierExpiresAt else { return false }
        return deadline < Date()
    }
}

struct FamilyMember: Codable, Identifiable {
    let id: UUID
    let familyId: UUID
    let userId: UUID
    var role: UserRole
    var status: MemberStatus
    let invitedAt: Date?
    let joinedAt: Date?

    // Joined fields
    var user: AppUser?

    enum CodingKeys: String, CodingKey {
        case id, role, status
        case familyId = "family_id"
        case userId = "user_id"
        case invitedAt = "invited_at"
        case joinedAt = "joined_at"
        case user = "users"
    }
}
